cpp/src/arrow/io/file.cc

```diff
@@ -148,7 +153,11 @@ class OSFile {
     if (pos < 0) {
       return Status::Invalid("Invalid position");
     }
-    return internal::FileSeek(fd_, pos);
+    Status st = internal::FileSeek(fd_, pos);
+    if (st.ok()) {
```

Comment:
What happens if a ReatAt is interleaved right here.

Reply:
Right. This PR is really a best effort solution to warn the user that they're shooting themselves in the foot. Inherently, we would need some locking to avoid races. See what I wrote under the discussion tab.